patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch

```diff
@@ -71,6 +94,189 @@
++		/// Use this method to spawn a emote bubble
++		/// </summary>
++		/// <param name="emoticon">The emote ID of the emote that will spawn.</param>
++		/// <param name="bubbleAnchor">The <see cref="WorldUIAnchor"/> instance for the emote. You can use <code>new WorldUIAnchor(Entity)</code> to get the instance.</param>
++		/// <param name="time">How long this emote remains.</param>
++		/// <returns>The <see cref="WhoAmI"/> of this emote</returns>
+ 		public static int NewBubble(int emoticon, WorldUIAnchor bubbleAnchor, int time) {
+ 			EmoteBubble emoteBubble = new EmoteBubble(emoticon, bubbleAnchor, time);
+ 			emoteBubble.ID = AssignNewID();
+@@ -92,22 +_,43 @@
+ 			}
+ 
+ 			OnBubbleChange(emoteBubble.ID);
++
++			EmoteBubbleLoader.OnSpawn(emoteBubble);
++
+ 			return emoteBubble.ID;
+ 		}
+ 
++		/// <summary>
++		/// Use this method to make NPCs use a random emote based on the pick emote table.
++		/// </summary>
++		/// <param name="bubbleAnchor">The <see cref="WorldUIAnchor"/> instance for the emote. You can use <code>new WorldUIAnchor(NPC)</code> to get the instance.</param>
++		/// <param name="time">How long this emote remains.</param>
++		/// <param name="other">The <see cref="WorldUIAnchor"/> instance from the other side of the conversation.</param>
++		/// <returns>The <see cref="WhoAmI"/> of this emote</returns>
+ 		public static int NewBubbleNPC(WorldUIAnchor bubbleAnchor, int time, WorldUIAnchor other = null) {
+ 			EmoteBubble emoteBubble = new EmoteBubble(0, bubbleAnchor, time);
++			emoteBubble.PickNPCEmote(out bool useEmote, other);
++			if (!useEmote) {
++				return -1;
++			}
+ 			emoteBubble.ID = AssignNewID();
+ 			byID[emoteBubble.ID] = emoteBubble;
+-			emoteBubble.PickNPCEmote(other);
++
+ 			if (Main.netMode == 2) {
+ 				Tuple<int, int> tuple = SerializeNetAnchor(bubbleAnchor);
+ 				NetMessage.SendData(91, -1, -1, null, emoteBubble.ID, tuple.Item1, tuple.Item2, time, emoteBubble.emote, emoteBubble.metadata);
+ 			}
+ 
++			EmoteBubbleLoader.OnSpawn(emoteBubble);
++
+ 			return emoteBubble.ID;
+ 		}
+ 
++		/// <summary>
++		/// Try to find a NPC close enough (less than 200 pixels) to react to the emote sent by the player.
++		/// </summary>
++		/// <param name="emoteID"></param>
++		/// <param name="player"></param>
+ 		public static void CheckForNPCsToReactToEmoteBubble(int emoteID, Player player) {
+ 			for (int i = 0; i < 200; i++) {
+ 				NPC nPC = Main.npc[i];
+@@ -130,7 +_,12 @@
+ 		}
+ 
+ 		private void Update() {
++			lifeTime--;
++			if (!EmoteBubbleLoader.UpdateFrame(this)) {
++				return;
++			}
++			// TML: Fixed frameSpeed not being used, might be a bug?
+-			if (--lifeTime > 0 && ++frameCounter >= 8) {
++			if (lifeTime > 0 && ++frameCounter >= frameSpeed) {
+ 				frameCounter = 0;
+ 				if (++frame >= 2)
+ 					frame = 0;
+@@ -139,11 +_,12 @@
+ 
+ 		private void Draw(SpriteBatch sb) {
+ 			Texture2D value = TextureAssets.Extra[48].Value;
++			Texture2D bubbleTexture = TextureAssets.Extra[48].Value;
```

Comment:
Suggest `bubbleTexture = value`, could be moved down below, to where `ModEmoteBubble` kicks in